llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -16795,6 +16795,16 @@ bool AArch64TargetLowering::optimizeExtendOrTruncateConversion(
 
       DstTy = TruncDstType;
     }
+
+    // mul(zext(i8), sext) can be transformed into smull(zext, sext) when
+    // destination type is at least SrcWidth * 4, which is faster than using tbl
+    // instructions
```

Comment:
I think that should at most SrcWidth * 4?

```suggestion
    // mul(zext(i8), sext) can be transformed into smull(zext, sext) which performs one extend implicitly. If DstWidth is at most 4 * SrcWidth, at most one extra extend step is needed and using tbl is not profitable.
```

Reply:
Is it? I might not understand something, but If SrcWidth is 8, DstWidth should be 32 or 64, meaning at least `SrcWidth*4`. "At most" would mean 8, 16, 32.

Reply:
Right, I guess the main point is that the `tbl` lowering is profitable if the remaining extends require at least 2 more steps, so if DstWidth is 64, the remaining extends would be from i8 -> i32, for which `tbl` for which tbl would still be profitable?

It's a moot point because we don't emit `tbl` for extends to i64 (don't remember why there's that restriction), but would be good to have consistent reasoning across the function.

This actually reminded me that support for skipping `tbl` for various widening instructions was added in e97b8a7e3fb9d4bd270bb25bac9777d86dcbdaf3 with the same reasoning as mentioned above. 

Is it possible that AArch64TTI::isWideningInstruction does not recognize `smull` and hence the general logic doesn't trigger?

Reply:
I happened to be looking at something else yesterday where isWideningInstr should maybe have been returning true if the size was more than double, but I think it might need to be more precise than that. Part of the extend becomes free, not the entire operation, and it should be accounted for somewhere. Whether that means that isWideningInstruction should be true or not I didn't look into, I would guess it probably could so long as the extend cost was accounted for.

Reply:
I was looking at the cost of v2i64 mul in some cases. We need to be careful that they don't get scalarized again when we do reduce the costs.

Reply:
@fhahn, as far as I understand, `AArch64TTI::isWideningInstruction` returns true only if all ZExt/SExt gets removed, but in our case, only one step of widening can be skipped:

> // Return true if the legalized types have the same number of vector elements
> // and the destination element type size is twice that of the source type.

Our case can be called partially widening, but I am not sure it is a good function to create.